cmd/fs-v1-metadata.go

```diff
@@ -274,167 +259,106 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
-// if already present and migrates to newer version if necessary. Returns
-// the final format version.
 func initFormatFS(fsPath, fsUUID string) (err error) {
-	fsFormatPath := pathJoin(fsPath, minioMetaBucket, fsFormatJSONFile)
-
-	// fsFormatJSONFile - format.json file stored in minioMetaBucket(.minio.sys) directory.
-	lk, err := lock.LockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
-	if err != nil {
-		return traceError(err)
-	}
-	defer lk.Close()
-
-	var format = &formatConfigV1{}
-	_, err = format.ReadFrom(lk)
-	// For all unexpected errors, we return.
-	if err != nil && errorCause(err) != io.EOF {
-		return traceError(fmt.Errorf("Unable to load 'format.json', %s", err))
-	}
-
-	// If we couldn't read anything, The disk is unformatted.
-	if errorCause(err) == io.EOF {
-		err = errUnformattedDisk
-		format = newFSFormatV2()
-	} else {
-		// Validate loaded `format.json`.
-		err = checkFormatFS(format, fsFormatVersion)
-		if err != nil && err != errFSFormatOld {
-			return traceError(fmt.Errorf("Unable to validate 'format.json', %s", err))
-		}
-	}
-
-	// Disk is in old format migrate object metadata.
-	if err == errFSFormatOld {
-		if merr := migrateFSObject(fsPath, fsUUID); merr != nil {
-			return merr
-		}
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, formatConfigFile)
+
+	// Once the filesystem has initialized hold the read lock for
+	// the life time of the server. This is done to ensure that under
+	// shared backend mode for FS, remote servers do not migrate
+	// or cause changes on backend format.
+
+	// This loop validates format.json by holding a read lock and
+	// proceeds if disk unformatted to hold non-blocking WriteLock
+	// If for some reason non-blocking WriteLock fails and the error
+	// is lock.ErrAlreadyLocked i.e some other process is holding a
+	// lock we retry in the loop again.
+	var rlk *lock.RLockedFile
+	for {
+		// Validate the `format.json` for expected values.
+		rlk, err = checkLockedValidFormatFS(fsPath)
+		switch {
+		case err == nil:
+			// Holding a read lock ensures that any write lock operation
+			// is blocked if attempted in-turn avoiding corruption on
+			// the backend disk.
+			_ = rlk // Hold the lock on `format.json` until server dies.
+			return nil
+		case errorCause(err) == errUnformattedDisk:
+			// Attempt a write lock on formatConfigFile `format.json`
+			// file stored in minioMetaBucket(.minio.sys) directory.
+			var lk *lock.LockedFile
+			lk, err = lock.TryLockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
+			if err != nil {
+				// Existing write locks detected.
+				if err == lock.ErrAlreadyLocked {
+					// Lock already present, sleep and attempt again.
+					time.Sleep(100 * time.Millisecond)
+					continue
+				}
+
+				// Unexpected error, return.
+				return traceError(err)
+			}
 
-		// Initialize format v2.
-		format = newFSFormatV2()
-	}
+			// Write new format.
+			if err = writeFormatFS(lk); err != nil {
+				return err
+			}
```

Comment:
A comment here saying that we continue the loop to get a read-lock and then finally exit the loop would be helpful to readers.